File: py_code/专项/贪心/763. 划分字母区间.py
from typing import List
'''
字符串 S 由小写字母组成。
我们要把这个字符串划分为尽可能多的片段，同一字母最多出现在一个片段中。
返回一个表示每个字符串片段的长度的列表。 
'''
'''
用列表记录对应字符出现的位置，保留最后出现的位置
'''
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        hash=[0]*26
        for i in range(len(s)):
            hash[ord(s[i])-ord('a')]=i

        #print(hash)
        
        res,r,l=[],0,0
        for i in range(len(s)):
            r=max(r,hash[ord(s[i])-ord('a')])
            if r==i:
                res.append(r-l+1)
                l=i+1
        return res

File: js_code/专项/贪心算法/763. 划分字母区间.js
/**
 * @param {string} s
 * @return {number[]}
 */
`
字符串 S 由小写字母组成。我们要把这个字符串划分为尽可能多的片段，
同一字母最多出现在一个片段中。返回一个表示每个字符串片段的长度的列表。
`
`
用对象记录字符串每个字符最后出现的位置，
当该字符当前位置等于记录最大位置时，说明该位置为分割点

这里记录最大位置表示为在该位置之前，记录的最大的最后出现位置
`
 var partitionLabels = function(s) {

  let hash={}
  for(let i=0;i<s.length;i++) hash[s[i]]=i
  //用的是对象，键值不能重复，所以之前的位置会被后面的位置覆盖，最后只留下最后出现的位置
  //console.log(hash)
  let res=[],right=0,left=0
  for(let i=0;i<s.length;i++){
      right=Math.max(right,hash[s[i]])

      if(right===i){
          res.push(right-left+1)
          left=i+1
      }
  }
  return res
};